File: tools/agentstream/checklist/tracker_writer.go
// Atomic Authoritative Tracker writer and loop-back invalidation.
// [IMPL-TIED_CHECKLIST_GATE_ENFORCEMENT] [ARCH-TIED_CHECKLIST_GATE_ENFORCEMENT] [REQ-TIED_CHECKLIST_GATE_ENFORCEMENT]
// How: APPLY_TRACKER_DISPOSITION and INVALIDATE_TRACKER_DOWNSTREAM mutate Tracker state atomically.
package checklist

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

// TurnIdentity binds a receipt to one checklist turn for replay protection.
type TurnIdentity struct {
	TurnIndex int
	StepStub  string
	SessionID string
}

func (id TurnIdentity) String() string {
	return fmt.Sprintf("turn=%d:stub=%s:session=%s", id.TurnIndex, strings.TrimSpace(id.StepStub), strings.TrimSpace(id.SessionID))
}

// ApplyTrackerDisposition merges one validated receipt into exactly one Tracker step.
func ApplyTrackerDisposition(trackerPath string, receipt CompletionReceipt, identity TurnIdentity) error {
	tracker, err := LoadTrackerYAML(trackerPath)
	if err != nil {
		return fmt.Errorf("tracker_not_found: %w", err)
	}
	hash := ReceiptHash(receipt)
	if replay, ok := findStateHistoryReplay(tracker, identity.String(), hash); ok {
		if replay {
			return nil
		}
		return fmt.Errorf("conflicting_replay")
	}
	stepsRaw, ok := tracker["steps"].([]interface{})
	if !ok {
		return fmt.Errorf("malformed_tracker: missing steps")
	}
	updated := false
	for i, item := range stepsRaw {
		row, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		slug, _ := row["slug"].(string)
		if strings.TrimSpace(slug) != strings.TrimSpace(receipt.Slug) {
			continue
		}
		applyReceiptToRow(row, receipt)
		stepsRaw[i] = row
		updated = true
		break
	}
	if !updated {
		return fmt.Errorf("step_not_found: %q", receipt.Slug)
	}
	tracker["steps"] = stepsRaw
	if ee, ok := tracker["execution_evidence"].(map[string]interface{}); ok {
		ee["completed"] = deriveCompletedSlugs(stepsRaw)
	} else {
		tracker["execution_evidence"] = map[string]interface{}{
			"completed": deriveCompletedSlugs(stepsRaw),
		}
	}
	appendStateHistory(tracker, identity.String(), hash)
	return atomicWriteYAML(trackerPath, tracker)
}

func applyReceiptToRow(row map[string]interface{}, receipt CompletionReceipt) {
	clearStepEvidence(row)
	row["disposition"] = receipt.Disposition
	row["updated_at"] = time.Now().UTC().Format(time.RFC3339)
	switch receipt.Disposition {
	case "completed":
		row["evidence_refs"] = stringListToInterface(receipt.EvidenceRefs)
	case "not_applicable":
		row["policy"] = receipt.Policy
		row["rationale"] = receipt.Rationale
	case "waived":
		row["owner"] = receipt.Owner
		row["expiry"] = receipt.Expiry
		row["approval"] = receipt.Approval
		row["residual_risk"] = receipt.ResidualRisk
	}
}

func clearStepEvidence(row map[string]interface{}) {
	for _, key := range []string{
		"evidence_refs", "policy", "rationale", "owner", "expiry", "approval", "residual_risk",
		"gate_receipt", "gate_summary",
	} {
		delete(row, key)
	}
}

func deriveCompletedSlugs(steps []interface{}) []interface{} {
	out := make([]interface{}, 0)
	for _, item := range steps {
		row, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		disp, _ := row["disposition"].(string)
		if disp != "completed" {
			continue
		}
		slug, _ := row["slug"].(string)
		slug = strings.TrimSpace(slug)
		if slug != "" {
			out = append(out, slug)
		}
	}
	return out
}

func appendStateHistory(tracker map[string]interface{}, turnIdentity, receiptHash string) {
	entry := map[string]interface{}{
		"turn_identity": turnIdentity,
		"receipt_hash":  receiptHash,
		"updated_at":    time.Now().UTC().Format(time.RFC3339),
	}
	history, _ := tracker["state_history"].([]interface{})
	history = append(history, entry)
	tracker["state_history"] = history
}

func findStateHistoryReplay(tracker map[string]interface{}, turnIdentity, receiptHash string) (idempotent bool, found bool) {
	history, _ := tracker["state_history"].([]interface{})
	for _, item := range history {
		row, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		id, _ := row["turn_identity"].(string)
		if id != turnIdentity {
			continue
		}
		hash, _ := row["receipt_hash"].(string)
		return hash == receiptHash, true
	}
	return false, false
}

// InvalidateTrackerDownstream resets configured clear_slugs after a validated goto.
func InvalidateTrackerDownstream(trackerPath, definitionPath, gotoTarget string) ([]string, error) {
	data, err := os.ReadFile(definitionPath)
	if err != nil {
		return nil, err
	}
	var doc yamlDoc
	if err := yaml.Unmarshal(data, &doc); err != nil {
		return nil, err
	}
	lb, ok := doc.LoopBack[strings.TrimSpace(gotoTarget)]
	if !ok || len(lb.ClearSlugs) == 0 {
		return nil, fmt.Errorf("missing_clear_target: %q", gotoTarget)
	}
	tracker, err := LoadTrackerYAML(trackerPath)
	if err != nil {
		return nil, err
	}
	stepsRaw, ok := tracker["steps"].([]interface{})
	if !ok {
		return nil, fmt.Errorf("malformed_tracker: missing steps")
	}
	indexBySlug := map[string]int{}
	for i, item := range stepsRaw {
		row, ok := item.(map[string]interface{})
		if !ok {
			continue
		}
		slug, _ := row["slug"].(string)
		slug = strings.TrimSpace(slug)
		if slug != "" {
			indexBySlug[slug] = i
		}
	}
	for _, slug := range lb.ClearSlugs {
		slug = strings.TrimSpace(slug)
		idx, ok := indexBySlug[slug]
		if !ok {
			return nil, fmt.Errorf("missing_state_row: %q", slug)
		}
		row, ok := stepsRaw[idx].(map[string]interface{})
		if !ok {
			return nil, fmt.Errorf("missing_state_row: %q", slug)
		}
		clearStepEvidence(row)
		row["disposition"] = "pending"
		delete(row, "updated_at")
		stepsRaw[idx] = row
	}
	tracker["steps"] = stepsRaw
	if ee, ok := tracker["execution_evidence"].(map[string]interface{}); ok {
		ee["completed"] = deriveCompletedSlugs(stepsRaw)
		clearCloseOutGateSummaries(ee, lb.ClearSlugs)
	}
	return append([]string(nil), lb.ClearSlugs...), atomicWriteYAML(trackerPath, tracker)
}

func clearCloseOutGateSummaries(ee map[string]interface{}, slugs []string) {
	closeOut, ok := ee["close_out_evidence"].(map[string]interface{})
	if !ok {
		return
	}
	gates, ok := closeOut["gates"].(map[string]interface{})
	if !ok {
		return
	}
	for _, slug := range slugs {
		delete(gates, slug)
	}
}

func stringListToInterface(items []string) []interface{} {
	out := make([]interface{}, 0, len(items))
	for _, item := range items {
		out = append(out, item)
	}
	return out
}

// atomicWriteYAML writes YAML via same-directory temp file and rename.
func atomicWriteYAML(path string, doc interface{}) error {
	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("write_failure: %w", err)
	}
	body, err := yaml.Marshal(doc)
	if err != nil {
		return fmt.Errorf("write_failure: %w", err)
	}
	tmp, err := os.CreateTemp(dir, ".tracker-*.yaml.tmp")
	if err != nil {
		return fmt.Errorf("write_failure: %w", err)
	}
	tmpPath := tmp.Name()
	defer func() {
		_ = os.Remove(tmpPath)
	}()
	if _, err := tmp.Write(body); err != nil {
		_ = tmp.Close()
		return fmt.Errorf("write_failure: %w", err)
	}
	if err := tmp.Sync(); err != nil {
		_ = tmp.Close()
		return fmt.Errorf("write_failure: %w", err)
	}
	if err := tmp.Close(); err != nil {
		return fmt.Errorf("write_failure: %w", err)
	}
	if err := os.Rename(tmpPath, path); err != nil {
		return fmt.Errorf("write_failure: %w", err)
	}
	return nil
}

// DerivedCompletedFromTracker returns execution_evidence.completed from step dispositions only.
func DerivedCompletedFromTracker(tracker map[string]interface{}) []string {
	stepsRaw, ok := tracker["steps"].([]interface{})
	if !ok {
		return nil
	}
	out := make([]string, 0)
	for _, item := range deriveCompletedSlugs(stepsRaw) {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

// ExecutionEvidenceCompletedOverride ignored when steps carry authoritative dispositions.
func ExecutionEvidenceCompletedOverride(tracker map[string]interface{}, override []string) bool {
	derived := DerivedCompletedFromTracker(tracker)
	if len(override) == 0 {
		return false
	}
	if len(derived) != len(override) {
		return true
	}
	seen := map[string]struct{}{}
	for _, slug := range derived {
		seen[slug] = struct{}{}
	}
	for _, slug := range override {
		if _, ok := seen[slug]; !ok {
			return true
		}
	}
	return false
}

File: mcp-server/packages/agentstream/src/tracker-writer.ts
/**
 * [IMPL-TIED_CHECKLIST_GATE_ENFORCEMENT] [ARCH-TIED_CHECKLIST_GATE_ENFORCEMENT] [REQ-TIED_CHECKLIST_GATE_ENFORCEMENT]
 * Tracker disposition apply + loop-back invalidation (Go checklist/tracker_writer.go parity).
 */
import fs from "node:fs";
import path from "node:path";

import yaml from "js-yaml";

import { findRepoRootFromPath } from "./repo-root.js";
import { atomicWriteYaml } from "./tracker-atomic-write.js";
import type { TrackerCompletionReceipt } from "./tracker-receipt.js";
import { loadTrackerYaml, type TrackerDoc } from "./tracker-yaml.js";

export type TurnIdentity = {
  turnIndex: number;
  stepStub: string;
  sessionId: string;
};

export function turnIdentityString(id: TurnIdentity): string {
  return `turn=${id.turnIndex}:stub=${id.stepStub.trim()}:session=${id.sessionId.trim()}`;
}

type LoopBackDoc = {
  loop_back_clearance?: Record<string, { clear_slugs?: string[] }>;
};

function clearStepEvidence(row: Record<string, unknown>): void {
  for (const key of [
    "evidence_refs",
    "policy",
    "rationale",
    "owner",
    "expiry",
    "approval",
    "residual_risk",
    "gate_receipt",
    "gate_summary",
  ]) {
    delete row[key];
  }
}

function evidenceRefsToInterface(
  refs: TrackerCompletionReceipt["evidence_refs"],
): unknown[] {
  if (!refs) {
    return [];
  }
  return [...refs];
}

function deriveCompletedSlugs(steps: unknown[]): unknown[] {
  const out: unknown[] = [];
  for (const item of steps) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      continue;
    }
    const row = item as Record<string, unknown>;
    if (String(row.disposition ?? "") !== "completed") {
      continue;
    }
    const slug = String(row.slug ?? "").trim();
    if (slug !== "") {
      out.push(slug);
    }
  }
  return out;
}

function findStateHistoryReplay(
  tracker: TrackerDoc,
  turnIdentity: string,
  receiptHash: string,
): { found: boolean; idempotent: boolean } {
  const history = tracker.state_history;
  if (!Array.isArray(history)) {
    return { found: false, idempotent: false };
  }
  for (const item of history) {
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      continue;
    }
    const row = item as Record<string, unknown>;
    if (String(row.turn_identity ?? "") !== turnIdentity) {
      continue;
    }
    return {
      found: true,
      idempotent: String(row.receipt_hash ?? "") === receiptHash,
    };
  }
  return { found: false, idempotent: false };
}

function appendStateHistory(
  tracker: Record<string, unknown>,
  turnIdentity: string,
  hash: string,
): void {
  const entry = {
    turn_identity: turnIdentity,
    receipt_hash: hash,
    updated_at: new Date().toISOString(),
  };
  const history = Array.isArray(tracker.state_history)
    ? [...tracker.state_history]
    : [];
  history.push(entry);
  tracker.state_history = history;
}

function applyReceiptToRow(
  row: Record<string, unknown>,
  receipt: TrackerCompletionReceipt,
): void {
  clearStepEvidence(row);
  row.disposition = receipt.disposition;
  row.updated_at = new Date().toISOString();
  switch (receipt.disposition) {
    case "completed":
      row.evidence_refs = evidenceRefsToInterface(receipt.evidence_refs);
      break;
    case "not_applicable":
      row.policy = receipt.policy;
      row.rationale = receipt.rationale;
      break;
    case "waived":
      row.owner = receipt.owner;
      row.expiry = receipt.expiry;
      row.approval = receipt.approval;
      row.residual_risk = receipt.residual_risk;
      break;
    default:
      break;
  }
}

function loadChecklistSlugsFromFile(checklistPath: string): Set<string> {
  const raw = fs.readFileSync(checklistPath, "utf8");
  const doc = yaml.load(raw) as {
    steps?: { slug?: string }[];
    sub_procedures?: { slug?: string }[];
  };
  const out = new Set<string>();
  for (const step of doc.steps ?? []) {
    const slug = String(step.slug ?? "").trim();
    if (slug !== "") {
      out.add(slug);
    }
  }
  for (const sub of doc.sub_procedures ?? []) {
    const slug = String(sub.slug ?? "").trim();
    if (slug !== "") {
      out.add(slug);
    }
  }
  return out;
}

let cachedCanonicalRoot = "";
let cachedCanonicalSlugs: Set<string> | null = null;

function slugRegistryForTracker(trackerPath: string): Set<string> {
  try {
    const tracker = loadTrackerYaml(trackerPath);
    const src = String(tracker.source_document ?? "").trim();
    if (src !== "") {
      const slugs = loadChecklistSlugsFromFile(src);
      if (slugs.size > 0) {
        return slugs;
      }
    }
  } catch {
    /* fall through */
  }
  const repoRoot = findRepoRootFromPath(trackerPath);
  if (repoRoot === "") {
    throw new Error("registry_unavailable");
  }
  if (cachedCanonicalSlugs !== null && cachedCanonicalRoot === repoRoot) {
    return cachedCanonicalSlugs;
  }
  const checklistPath = path.join(
    repoRoot,
    "tied",
    "docs",
    "agent-req-implementation-checklist.yaml",
  );
  const slugs = loadChecklistSlugsFromFile(checklistPath);
  cachedCanonicalRoot = repoRoot;
  cachedCanonicalSlugs = slugs;
  return slugs;
}

export function validateCanonicalSlug(trackerPath: string, slug: string): void {
  const s = slug.trim();
  if (s === "") {
    throw new Error("invalid_slug:empty");
  }
  let slugs: Set<string>;
  try {
    slugs = slugRegistryForTracker(trackerPath);
  } catch {
    throw new Error("invalid_slug:registry_unavailable");
  }
  if (!slugs.has(s)) {
    throw new Error(`invalid_slug:${s}`);
  }
}

export function applyTrackerDisposition(
  trackerPath: string,
  receipt: TrackerCompletionReceipt,
  identity: TurnIdentity,
  receiptHashHex: string,
): void {
  validateCanonicalSlug(trackerPath, receipt.slug);
  const tracker = loadTrackerYaml(trackerPath) as Record<string, unknown>;
  const turnId = turnIdentityString(identity);
  const replay = findStateHistoryReplay(
    tracker as TrackerDoc,
    turnId,
    receiptHashHex,
  );
  if (replay.found) {
    if (replay.idempotent) {
      return;
    }
    throw new Error("conflicting_replay");
  }
  const stepsRaw = tracker.steps;
  if (!Array.isArray(stepsRaw)) {
    throw new Error("malformed_tracker: missing steps");
  }
  let updated = false;
  for (let i = 0; i < stepsRaw.length; i++) {
    const item = stepsRaw[i];
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      continue;
    }
    const row = item as Record<string, unknown>;
    if (String(row.slug ?? "").trim() !== String(receipt.slug).trim()) {
      continue;
    }
    applyReceiptToRow(row, receipt);
    stepsRaw[i] = row;
    updated = true;
    break;
  }
  if (!updated) {
    throw new Error(`step_not_found: ${JSON.stringify(receipt.slug)}`);
  }
  tracker.steps = stepsRaw;
  const ee =
    tracker.execution_evidence !== null &&
    typeof tracker.execution_evidence === "object" &&
    !Array.isArray(tracker.execution_evidence)
      ? (tracker.execution_evidence as Record<string, unknown>)
      : {};
  ee.completed = deriveCompletedSlugs(stepsRaw);
  tracker.execution_evidence = ee;
  appendStateHistory(tracker, turnId, receiptHashHex);
  atomicWriteYaml(trackerPath, tracker);
}

function clearCloseOutGateSummaries(
  ee: Record<string, unknown>,
  slugs: string[],
): void {
  const closeOut = ee.close_out_evidence;
  if (closeOut === null || typeof closeOut !== "object" || Array.isArray(closeOut)) {
    return;
  }
  const gates = (closeOut as Record<string, unknown>).gates;
  if (gates === null || typeof gates !== "object" || Array.isArray(gates)) {
    return;
  }
  const gateMap = gates as Record<string, unknown>;
  for (const slug of slugs) {
    delete gateMap[slug];
  }
}

export function invalidateTrackerDownstream(
  trackerPath: string,
  definitionPath: string,
  gotoTarget: string,
): string[] {
  const data = fs.readFileSync(definitionPath, "utf8");
  const doc = yaml.load(data) as LoopBackDoc;
  const lb = doc.loop_back_clearance?.[gotoTarget.trim()];
  if (!lb?.clear_slugs || lb.clear_slugs.length === 0) {
    throw new Error(`missing_clear_target: ${JSON.stringify(gotoTarget)}`);
  }
  const tracker = loadTrackerYaml(trackerPath) as Record<string, unknown>;
  const stepsRaw = tracker.steps;
  if (!Array.isArray(stepsRaw)) {
    throw new Error("malformed_tracker: missing steps");
  }
  const indexBySlug = new Map<string, number>();
  for (let i = 0; i < stepsRaw.length; i++) {
    const item = stepsRaw[i];
    if (item === null || typeof item !== "object" || Array.isArray(item)) {
      continue;
    }
    const slug = String((item as Record<string, unknown>).slug ?? "").trim();
    if (slug !== "") {
      indexBySlug.set(slug, i);
    }
  }
  for (const slug of lb.clear_slugs) {
    const s = slug.trim();
    const idx = indexBySlug.get(s);
    if (idx === undefined) {
      throw new Error(`missing_state_row: ${JSON.stringify(s)}`);
    }
    const row = stepsRaw[idx] as Record<string, unknown>;
    clearStepEvidence(row);
    row.disposition = "pending";
    delete row.updated_at;
    stepsRaw[idx] = row;
  }
  tracker.steps = stepsRaw;
  const ee =
    tracker.execution_evidence !== null &&
    typeof tracker.execution_evidence === "object" &&
    !Array.isArray(tracker.execution_evidence)
      ? (tracker.execution_evidence as Record<string, unknown>)
      : undefined;
  if (ee) {
    ee.completed = deriveCompletedSlugs(stepsRaw);
    clearCloseOutGateSummaries(ee, lb.clear_slugs);
  }
  atomicWriteYaml(trackerPath, tracker);
  return [...lb.clear_slugs];
}
